java/rocksjni/comparatorjnicallback.cc

```diff
@@ -202,139 +429,214 @@ void BaseComparatorJniCallback::FindShortSuccessor(
+            m_jbytebuffer_clazz));
+        if (jtl_buf == nullptr) {
+          // exception thrown: OutOfMemoryError
+          return nullptr;
+        }
+        tlb = new ThreadLocalBuf(m_jvm, m_options->direct_buffer, jtl_buf);
+        tl_buf->Reset(tlb);
+      }
+      return ReuseBuffer(env, src, tlb->jbuf);
+    } else {
 
-  m_jSliceB = env->NewGlobalRef(DirectSliceJni::construct0(env));
-  if(m_jSliceB == nullptr) {
-    // exception thrown: OutOfMemoryError
-    return;
-  }
+      // reuse class member buffer
+      return ReuseBuffer(env, src, jreuse_buffer);
+    }
+  } else {
 
-  m_jSliceLimit = env->NewGlobalRef(DirectSliceJni::construct0(env));
-  if(m_jSliceLimit == nullptr) {
-    // exception thrown: OutOfMemoryError
-    return;
+    // new buffer
+    return NewBuffer(env, src);
   }
 }
 
-DirectComparatorJniCallback::~DirectComparatorJniCallback() {
-  jboolean attached_thread = JNI_FALSE;
-  JNIEnv* env = getJniEnv(&attached_thread);
-  assert(env != nullptr);
-
-  if(m_jSliceA != nullptr) {
-    env->DeleteGlobalRef(m_jSliceA);
-  }
-
-  if(m_jSliceB != nullptr) {
-    env->DeleteGlobalRef(m_jSliceB);
+jobject ComparatorJniCallback::ReuseBuffer(
+    JNIEnv* env, const Slice& src, jobject jreuse_buffer) const {
+  // we can reuse the buffer
+  if (m_options->direct_buffer) {
+    // copy into direct buffer
+    void* buf = env->GetDirectBufferAddress(jreuse_buffer);
+    if (buf == nullptr) {
+      // either memory region is undefined, given object is not a direct java.nio.Buffer, or JNI access to direct buffers is not supported by this virtual machine.
+      rocksdb::RocksDBExceptionJni::ThrowNew(env, "Unable to get Direct Buffer Address");
+      return nullptr;
+    }
+    memcpy(buf, src.data(), src.size());
+  } else {
+    // copy into non-direct buffer
+    const jbyteArray jarray = ByteBufferJni::array(env, jreuse_buffer,
+        m_jbytebuffer_clazz);
+    if (jarray == nullptr) {
+      // exception occurred
+      return nullptr;
+    }
+    env->SetByteArrayRegion(jarray, 0, static_cast<jsize>(src.size()),
+        const_cast<jbyte*>(reinterpret_cast<const jbyte*>(src.data())));
+    if (env->ExceptionCheck()) {
+      // exception occurred
+      env->DeleteLocalRef(jarray);
+      return nullptr;
+    }
+    env->DeleteLocalRef(jarray);
   }
+  return jreuse_buffer;
+}
 
-  if(m_jSliceLimit != nullptr) {
-    env->DeleteGlobalRef(m_jSliceLimit);
+jobject ComparatorJniCallback::NewBuffer(JNIEnv* env, const Slice& src) const {
+  // we need a new buffer
+  jobject jbuf = ByteBufferJni::constructWith(env, m_options->direct_buffer,
+      src.data(), src.size(), m_jbytebuffer_clazz);
+  if (jbuf == nullptr) {
+    // exception occurred
+    return nullptr;
   }
+  return jbuf;
+}
 
-  releaseJniEnv(attached_thread);
+void ComparatorJniCallback::DeleteBuffer(JNIEnv* env, jobject jbuffer) const {
+  // if direct then free delete the underlying array too
+  // if (m_options->direct_buffer) {
```

Comment:
What's the deal with this commented out code?

Reply:
Its not needed I will remove it